src/Temporalio/Bridge/src/client.rs

```diff
@@ -0,0 +1,453 @@
+                        .alloc_utf8(&format!("Connection failed: {}", err))
+                        .into_raw(),
+                );
+            },
+        }
+    });
+}
+
+#[no_mangle]
+pub extern "C" fn client_free(client: *mut Client) {
+    unsafe {
+        let _ = Box::from_raw(client);
+    }
+}
+
+#[repr(C)]
+pub struct RpcCallOptions {
+    service: RpcService,
+    rpc: ByteArrayRef,
+    req: ByteArrayRef,
+    retry: bool,
+    /// Metadata is <key1>\n<value1>\n<key2>\n<value2>. Metadata keys or
+    /// values cannot contain a newline within.
+    metadata: ByteArrayRef,
+    /// 0 means no timeout
+    timeout_millis: u32,
+    cancellation_token: *const CancellationToken,
+}
+
+// Expected to outlive all async calls that use it
+unsafe impl Send for RpcCallOptions {}
+unsafe impl Sync for RpcCallOptions {}
+
+#[repr(C)]
+pub enum RpcService {
+    Workflow = 1,
+    Operator,
+    Test,
+    Health,
+}
+
+/// If success or fail are not null, they must be manually freed when done.
+type ClientRpcCallCallback = unsafe extern "C" fn(
+    user_data: *mut libc::c_void,
+    success: *const ByteArray,
+    fail: *const ByteArray,
+);
+
+macro_rules! service_call {
+    ($service_fn:ident, $client:ident, $options:ident, $cancel_token:ident) => {{
+        let call_future = $service_fn(&$client.core, &$options);
+        if let Some(cancel_token) = $cancel_token {
+            tokio::select! {
+                _ = cancel_token.cancelled() => Err(anyhow::anyhow!("Cancelled")),
+                v = call_future => v,
+            }
+        } else {
+            call_future.await
+        }
+    }};
+}
+
+/// Client, options, and user data must live through callback.
+#[no_mangle]
+pub extern "C" fn client_rpc_call(
+    client: *mut Client,
+    options: *const RpcCallOptions,
+    user_data: *mut libc::c_void,
+    callback: ClientRpcCallCallback,
+) {
+    let client = unsafe { &*client };
+    let options = unsafe { &*options };
+    let cancel_token = unsafe { options.cancellation_token.as_ref() }.map(|v| v.token.clone());
+    let user_data = UserDataHandle(user_data);
+    client.runtime.core.tokio_handle().spawn(async move {
+        let res = match options.service {
+            RpcService::Workflow => {
+                service_call!(call_workflow_service, client, options, cancel_token)
+            }
+            RpcService::Operator => {
+                service_call!(call_operator_service, client, options, cancel_token)
+            }
+            RpcService::Test => service_call!(call_test_service, client, options, cancel_token),
+            RpcService::Health => service_call!(call_health_service, client, options, cancel_token),
+        };
+        let (success, fail) = match res {
+            Ok(b) => (ByteArray::from_vec(b).into_raw(), std::ptr::null_mut()),
+            Err(err) => (
+                std::ptr::null_mut(),
+                ByteArray::from_utf8(format!("{}", err)).into_raw(),
```

Comment:
```suggestion
                client.runtime.alloc_utf8(format!("{}", err)).into_raw(),
```

Even though we're not using a shared buffer pool currently, we _could_ one day, so we need to centralize allocations (I forgot this one and the `ByteArray::from_vec` above)